File: android/src/main/java/com/audiobrowser/cast/CastReSignBudget.kt
package com.audiobrowser.cast

/**
 * Pure per-item bookkeeping for the bounded reactive re-sign of stale Cast queue URLs (see ADR
 * 0003): caps attempts per item id and dedups re-entrant triggers while a re-sign is in flight, so
 * a genuinely dead stream surfaces a real error instead of looping forever (mirrors
 * `StuckRecoveryPolicy`'s philosophy).
 *
 * Lives in **main** (no Cast SDK) so it compiles in the default build and is unit-testable. The
 * `cast` sourceset's `CastReSign` keeps the SDK resolve / `queueUpdateItems` calls and uses this for
 * the bookkeeping. Not thread-safe — callers confine to the main (RemoteMediaClient callback)
 * thread.
 */
class CastReSignBudget(private val maxAttemptsPerItem: Int = 3) {
  private val attemptsByItemKey = mutableMapOf<String, Int>()
  private val inFlight = mutableSetOf<String>()

  /**
   * Decides whether to dispatch a re-sign for [key] and, when yes, claims an attempt + marks it
   * in-flight (so re-entrant idle+error callbacks for the same item don't each burn the budget).
   *
   * - [Decision.ATTEMPT] — dispatch; remember to call [markDone] when the dispatch settles.
   * - [Decision.IN_FLIGHT] — a re-sign for this item is already running; do nothing (caller should
   *   NOT surface an error yet).
   * - [Decision.EXHAUSTED] — the per-item cap is spent; caller surfaces the real error.
   */
  fun shouldAttempt(key: String): Decision {
    if (key in inFlight) return Decision.IN_FLIGHT
    val attempts = attemptsByItemKey.getOrDefault(key, 0)
    if (attempts >= maxAttemptsPerItem) return Decision.EXHAUSTED
    attemptsByItemKey[key] = attempts + 1
    inFlight.add(key)
    return Decision.ATTEMPT
  }

  /** Clears the in-flight mark for [key] once its dispatched re-sign settles (success or failure). */
  fun markDone(key: String) {
    inFlight.remove(key)
  }

  /** Resets all counters — call on a fresh queue load / new Cast session. */
  fun reset() {
    attemptsByItemKey.clear()
    inFlight.clear()
  }

  enum class Decision {
    ATTEMPT,
    IN_FLIGHT,
    EXHAUSTED,
  }
}

File: ios/Cast/CastReSignBudget.swift
import Foundation

/// Pure per-item bookkeeping for the bounded, reactive re-signing of stale Cast
/// queue URLs (see ADR-0003): caps attempts per receiver item id and dedups
/// re-entrant triggers while a re-sign is in flight, so a genuinely dead stream
/// surfaces a real error instead of looping forever.
///
/// **Cast-SDK-free and UNGATED on purpose** — it touches only `UInt` item ids and
/// so compiles in the default (Cast-disabled) build and is unit-testable from the
/// SPM test target (mirrors Android's `CastReSignBudget` in the `main` sourceset).
/// The gated `CastReSign` keeps the GCK resolve / `queueUpdate` calls and delegates
/// all bookkeeping here. A value type confined to the main actor by its single
/// owner; not itself thread-safe.
struct CastReSignBudget {
  /// The outcome of asking whether to re-sign a failing item.
  enum Decision {
    /// Dispatch a re-sign; an attempt has been claimed + marked in flight. Call
    /// `markDone` when the dispatch settles.
    case attempt
    /// A re-sign for this item is already running — do nothing (the caller must
    /// NOT surface a terminal error yet).
    case inFlight
    /// The per-item cap is spent — the caller surfaces the real error.
    case exhausted
  }

  private let maxAttemptsPerItem: Int
  private var attemptsByItem: [UInt: Int] = [:]
  private var inFlight: Set<UInt> = []

  init(maxAttemptsPerItem: Int = 3) {
    self.maxAttemptsPerItem = maxAttemptsPerItem
  }

  /// Decides whether to dispatch a re-sign for `itemID` and, when yes, claims an
  /// attempt + marks it in flight (so re-entrant idle+error callbacks for the same
  /// item don't each burn the budget).
  mutating func shouldAttempt(_ itemID: UInt) -> Decision {
    if inFlight.contains(itemID) { return .inFlight }
    let used = attemptsByItem[itemID] ?? 0
    if used >= maxAttemptsPerItem { return .exhausted }
    attemptsByItem[itemID] = used + 1
    inFlight.insert(itemID)
    return .attempt
  }

  /// Whether a re-sign for `itemID` is currently dispatched and not yet settled.
  /// Lets a caller short-circuit a repeat trigger before doing any other work.
  func isInFlight(_ itemID: UInt) -> Bool {
    inFlight.contains(itemID)
  }

  /// Clears the in-flight mark for `itemID` once its dispatched re-sign settles
  /// (success or failure), so a later genuine expiry can be re-signed again.
  mutating func markDone(_ itemID: UInt) {
    inFlight.remove(itemID)
  }

  /// Resets all counters — call on a fresh queue load / new Cast session.
  mutating func reset() {
    attemptsByItem.removeAll()
    inFlight.removeAll()
  }
}
